packages/grid/x-grid/src/tests/rows.XGrid.test.tsx

```diff
@@ -249,4 +251,97 @@ describe('<XGrid /> - Rows ', () => {
       expect(getColumnValues()).to.deep.equal(['Apple', 'Atari']);
     });
   });
+
+  describe('virtualization', () => {
+    let apiRef: GridApiRef;
+    const TestCaseVirtualization = (
+      props: Partial<XGridProps> & { nbRows?: number; nbCols?: number; height?: number },
+    ) => {
+      apiRef = useGridApiRef();
+      const data = useData(props.nbRows || 100, props.nbCols || 10);
+
+      return (
+        <div style={{ width: 300, height: props.height || 300 }}>
+          <XGrid apiRef={apiRef} columns={data.columns} rows={data.rows} {...props} />
+        </div>
+      );
+    };
+
+    it('Rows should not be virtualized when the number of rows fit in the viewport', () => {
+      const headerHeight = 50;
+      const rowHeight = 50;
+      const maxRowsNotVirtualised = (300 - headerHeight) / rowHeight;
+      render(
+        <TestCaseVirtualization
+          nbRows={maxRowsNotVirtualised}
+          hideFooter
+          headerHeight={headerHeight}
+          rowHeight={rowHeight}
+        />,
+      );
+
+      const isVirtualized = apiRef!.current!.getState().containerSizes!.isVirtualized;
+      expect(isVirtualized).to.equal(false);
+    });
+
+    it('Rows should be virtualized when at least 2 rows are outside the viewport', () => {
+      const headerHeight = 50;
+      const rowHeight = 50;
+      const maxRowsNotVirtualised = (300 - headerHeight) / rowHeight;
+      render(
+        <TestCaseVirtualization
+          nbRows={maxRowsNotVirtualised + 1}
+          hideFooter
+          headerHeight={headerHeight}
+          rowHeight={rowHeight}
+        />,
+      );
+
+      let isVirtualized = apiRef!.current!.getState().containerSizes!.isVirtualized;
+      expect(isVirtualized).to.equal(false);
+
+      render(
+        <TestCaseVirtualization
+          nbRows={maxRowsNotVirtualised + 2}
+          hideFooter
+          headerHeight={headerHeight}
+          rowHeight={rowHeight}
+        />,
+      );
+
+      isVirtualized = apiRef!.current!.getState().containerSizes!.isVirtualized;
+      expect(isVirtualized).to.equal(true);
+    });
+
+    it('should render last row when scrolling to the bottom', () => {
+      const { container } = render(<TestCaseVirtualization nbRows={996} hideFooter height={600} />);
+      const totalHeight = apiRef!.current!.getState().containerSizes?.totalSizes.height;
+
+      const gridWindow = container.querySelector('.MuiDataGrid-window');
+      const renderingZone = container.querySelector('.MuiDataGrid-renderingZone');
+      gridWindow.scrollTop = totalHeight;
```

Comment:
How about we use the same approach as @DanailH https://github.com/mui-org/material-ui-x/blob/76948930e3a9fa6d0d03fe60c54b60d3a60e2202/packages/grid/x-grid/src/tests/events.XGrid.test.tsx#L269

But to hedge any potential issues with (12345 isn't enough) 

```suggestion
      gridWindow.scrollTop = 10e6;
```

This way, we can refactor the internal of the grid without having to care about `apiRef!.current!.getState().containerSizes?.totalSizes.height`. We scroll to the bottom, end of the story.

Reply:
Yes but if you put a random value you're not testing anything anymore. What I test here is that the calculated value corresponds to the end of the page

Reply:
> test here is that the calculated value corresponds to the end of the page

Then we can rewrite the expect to explicitly test it. It's not obvious. You could for instance assert that the scroll height is equal or set a really high scroll top and assert that the browsers max it out to the correct value?

Reply:
I don't understand why it's confusing, not sure about what you mean, how it will be a better test?

I assign totalHeight to the scroll top value so window goes down to the totalHeight value. 
Then I try to retrieve the last cell, so if the virtualisation or the totalHeight is wrong then I won't find it.
Finally I assert that the rendering zone has the expected transform value.

Reply:
I propose that we break`gridWindow.scrollTop = totalHeight;` down. As you said, this line serves two purposes:

1. It's meant to scroll to the bottom of the scroll container, to make sure that even the last cell is rendered
2. It's meant to make sure state.totalHeight is actually the total height.

For 1. I would propose:

```jsx
gridWindow.scrollTop = 10e6; // scroll to the bottom
```

no ambiguity possible about the objective 

For 2. I would propose:

```jsx
expect(gridWindow.scrollHeight).to.equal(totalHeight);
```

so we can also detect if `totalHeight` is too tall (no only too small). When the code assigns the `element.scrollTop` property, the browser do `Math.min(element.scrollHeight, new value)`.